### ISSUE TITLE
new rule: prefer-import-tag

### ISSUE THREAD

**Issue by boneskull** [gajus/eslint-plugin-jsdoc#1314]
## Motivation

I have a codebase littered with inline type-level imports (e.g., `@type {import('foo').bar}`).  I'd like to be able to put a halt to this madness and use [import tags](https://devblogs.microsoft.com/typescript/announcing-typescript-5-5/#the-jsdoc-@import-tag) instead.  I would _especially_ like this to be a fixable rule.

## Current behavior

It's the wild west out there

## Desired behavior

A rule which could raise an error if an inline type import was detected.  Default behavior of this rule, when enabled, would be to always prefer import tags over inline imports.

Highly desired, but not required for a first pass:

- automatically fixable (though this may be difficult or dangerous, because it will introduce symbols into the module namespace [insofar as TS is concerned] and symbols can conflict)

Optionally:

- inline imports could be preferred over import tags
- exceptions based on some criteria (e.g., imports from `foo` are an exception to the configured behavior)

## Alternatives considered

- While TS itself supports JSDoc, `typescript-eslint` does not and will not support JSDoc.
- I thought about writing a codemod but that seems too complicated.

* * *

I am unsure how feasible this is or how difficult an implementation would be, so I apologize in advance if it's ridiculous.

**Comment by brettz9** [comment]
You can currently get this without a fixer by using:

```js
'jsdoc/no-restricted-syntax': ['warn', {
  contexts: [
    {
      comment: 'JsdocBlock:has(JsdocTag:has(JsdocTypeImport))',
      context: 'any',
      message: 'Use @import tag over import() statements',
    },
  ],
}]
```

**Comment by JounQin** [comment]
@brettz9 `@import` is not reorganized currently:

```sh
  ⚠  jsdoc/check-values

     Bad @import tag


     packages/postcss-config/index.js:2:1
     1 | /**
   > 2 |  * @import {ConfigFn, ConfigPlugin} from 'postcss-load-config
       | ^
     3 |  */
     4 | 
     5 | import { NODE_ENV, __DEV__, __PROD__ } from '@pkgr/utils'

  ⚠  jsdoc/no-undefined-types

     The type 'ConfigFn' is undefined


     packages/postcss-config/index.js:16:1
     14 | /**
     15 |  *
   > 16 |  * @type {ConfigFn}
        | ^
     17 |  */
     18 | const config = ({
     19 |   advanced,

  ⚠  jsdoc/no-undefined-types

     The type 'ConfigPlugin' is undefined


     packages/postcss-config/index.js:30:1
     28 | } = {}) => {
     29 |   /**
   > 30 |    * @type {ConfigPlugin[]}
        | ^
     31 |    */
     32 |   const plugins = [
     33 |     presetEnv_(presetEnv),
```

**Comment by brettz9** [comment]
It looks like you are missing an ending apostrophe after `'postcss-load-config`.

**Comment by JounQin** [comment]
> It looks like you are missing an ending apostrophe after `'postcss-load-config`.

Holy... You're damn right, the error message is confusing.

**Comment by brettz9** [comment]
> > It looks like you are missing an ending apostrophe after `'postcss-load-config`.
> 
> Holy... You're damn right, the error message is confusing.

The `parse-imports` package we are using does not provide more specific feedback as to the type of error. We need to use one message to report any kind of `@import` error.

**Comment by brettz9** [comment]
Take a look at #1536 . It doesn't implement the features you mentioned as optional, and it doesn't add what perhaps some might want, optionally exporting as `@typedef`s instead of `@import` (potentially trying to be local (i.e., non-exported) `@typedef`s by inserting the typedef block above the comment block with the reference).

(While typedefs can conflict, an advantage of typedef export is that they can reference the whole expression like `@typedef {import('sth').TopLevelNamespace.Some.Long.Expression} LongExpression`, not just the `/** @import {TopLevelNamespace} from 'sth' */` which requires repeating "TopLevelNamespace.Some.Long.Expression".)
